File: rust/segment_tree.rs
fn init()
{
    let mut tree:Vec<usize> = vec![0;1 << 21];
    for i in 0..n { tree[i + (1 << 20)] = scanner.next().unwrap().unwrap() }
    for i in (1..=(1 << 20) - 1).rev() { tree[i] = tree[i * 2] + tree[i * 2 + 1] }
}

fn update(tree:&mut Vec<usize>, mut a:usize, b:usize)
{
    a += (1 << 20) - 1;

    tree[a] = b;

    a >>= 1;

    while a > 0 { tree[a] = tree[a * 2] + tree[a * 2 + 1]; a >>= 1 }
}

fn query (tree:&Vec<usize>, mut a:usize, mut b:usize) -> usize
{
    let mut res = 0;

    a += (1 << 20) - 1; b += (1 << 20) - 1;

    while a <= b
    {
        if a & 1 == 1 { res += tree[a]; a += 1 }
        if b & 1 == 0 { res += tree[b]; b -= 1 }

        a >>= 1; b >>= 1;
    }

    res
}

File: python3/segment_tree.py
def init(tree):
    N = len(tree) // 2
    
    for i in range(N - 1, 0, -1):
        tree[i] = max(tree[i * 2], tree[i * 2 + 1])
        
def update(tree, idx, val):
    idx += len(tree) // 2 - 1
    tree[idx] = val

    while idx > 1:
        idx //= 2
        tree[idx] = max(tree[idx * 2], tree[idx * 2 + 1])
        
def query(tree, idx1, idx2):
    res = 0

    idx1 += len(tree) // 2 - 1
    idx2 += len(tree) // 2 - 1

    while idx1 <= idx2:
        if idx1 % 2 == 1:
            res = max(res, tree[idx1])
            idx1 += 1

        if idx2 % 2 == 0:
            res = max(res, tree[idx2])
            idx2 -= 1

        idx1 //= 2
        idx2 //= 2

    return res
